core/domain/exp_services.py

```diff
@@ -1710,6 +1710,516 @@ def is_voiceover_change_list(change_list):
+                    changed_property = 'answer_groups'
+                # A condition to store the name of the properties changed
+                # in changed_properties dict.
+                state_name = change.state_name
+                if state_name in new_to_old_state_names:
+                    state_name = new_to_old_state_names[change.state_name]
+                if state_name in changed_translations:
+                    if (changed_property not in
+                            changed_translations[state_name]):
+                        changed_translations[state_name].append(
+                            changed_property)
+                else:
+                    changed_translations[state_name] = [changed_property]
+                if state_name in changed_properties:
+                    if ('written_translations' not in
+                            changed_properties[state_name]):
+                        changed_properties[state_name].append(
+                            'written_translations')
+                else:
+                    changed_properties[state_name] = [
+                        'written_translations']
+
+        # Old_to_new_state_names: dict. Dictionary mapping state names of
+        # prev_exp_version to the state names of current_exp_version.
+        # It doesn't include the name changes of added/deleted states.
+        old_to_new_state_names = {
+            value: key for key, value in new_to_old_state_names.items()
+        }
+
+        if len(added_state_names) > 0 or len(deleted_state_names) > 0:
+            # Here we will send the changelist, version, latest_version,
+            # and exploration to the admin, so that the conditions
+            # can be reviewed.
+            return False
+
+        changes_are_mergeable = False
+
+        # state_names_of_renamed_states: dict. Stores the changes in
+        # states names in change_list where the key is the state name in
+        # frontend version and the value is the renamed name from the
+        # change list if there is any rename state change.
+        state_names_of_renamed_states = {}
+        for change in change_list:
+            change_is_mergeable = False
+            if change.cmd == exp_domain.CMD_RENAME_STATE:
+                old_state_name = change.old_state_name
+                new_state_name = change.new_state_name
+                if old_state_name in state_names_of_renamed_states:
+                    state_names_of_renamed_states[new_state_name] = (
+                        state_names_of_renamed_states.pop(old_state_name))
+                else:
+                    state_names_of_renamed_states[new_state_name] = (
+                        old_state_name)
+                if (state_names_of_renamed_states[new_state_name] not in
+                        old_to_new_state_names):
+                    change_is_mergeable = True
+            elif change.cmd == exp_domain.CMD_EDIT_STATE_PROPERTY:
```

Comment:
I'm not reviewing the function below this line as I'm finding it hard to keep the context of every variable above and read another 250+ lines with that weightage. [I'll review the extra lines once the code structure is optimized.]

Reply:
Answered all the comments.